src/main/java/org/template/Algorithm.java

```diff
@@ -24,360 +32,846 @@
+      rankRDDs.add(new Tuple2<>(rankingFieldName, rankRdd));
+    }
 
-        // singleton list for propertiesRdd
-        ArrayList<JavaPairRDD<String, Map<String, JsonAST.JValue>>> pList = new ArrayList<>();
-        pList.add(propertiesRDD);
-        new URModel(
-                cooccurrenceCorrelators,
-                pList,
-                getRankingMapping(),
-                false,
-                sc).save(dateNames, esIndex, esType);
-        return new NullModel();
+    JavaPairRDD<String, Map<String, JsonAST.JValue>> acc = RDDUtils.getEmptyPairRDD(sc);
+    // TODO: Is functional [fold & map] more efficient than looping?
+    for (Tuple2<String, JavaPairRDD<String, Double>> t : rankRDDs) {
+      String fieldName = t._1();
+      JavaPairRDD<String, Double> rightRdd = t._2();
+      JavaPairRDD joined = acc.fullOuterJoin(rightRdd);
+      acc = joined.mapToPair(new RankFunction(fieldName));
     }
+    return acc;
+  }
+
+  public NullModel calcAll(SparkContext sc, PreparedData preparedData) {
+    return calcAll(sc, preparedData, true);
+  }
+
+  @Override
+  public NullModel train(SparkContext sc, PreparedData preparedData) {
+    if (this.recsModel.equals(RecsModel.All))
+      return this.calcAll(sc, preparedData);
+    else if (this.recsModel.equals(RecsModel.BF))
+      return this.calcPop(sc, preparedData);
+    else if (this.recsModel.equals(RecsModel.CF)) {
+      return this.calcAll(sc, preparedData, false);
+    } else {
+      throw new IllegalArgumentException(
+          String.format("| Bad algorithm param recsModel=[%s] in engine definition params, possibly a bad json value.  |Use one of the available parameter values (%s).",
+              this.recsModel, new RecsModel().toString())
+      );
+    }
+  }
+
+  public NullModel calcPop(SparkContext sc, PreparedData data) {
+    JavaPairRDD<String, Map<String, JsonAST.JValue>> fieldsRDD = data.getFieldsRDD();
+    JavaPairRDD<String, Map<String, JsonAST.JValue>> ranksRDD = getRanksRDD(fieldsRDD, sc);
+
+    JavaPairRDD<String, Map<String, JsonAST.JValue>> currentMetadataRDD = EsClient.getInstance().getRDD(esIndex, esType, sc);
+    JavaPairRDD<String, Map<String, JsonAST.JValue>> propertiesRDD = currentMetadataRDD.fullOuterJoin(ranksRDD).mapToPair(new CalcAllFunction());
+
+    // singleton list for propertiesRdd
+    ArrayList<JavaPairRDD<String, Map<String, JsonAST.JValue>>> pList = new ArrayList<>();
+    pList.add(fieldsRDD.cache());
+    pList.add(propertiesRDD.cache());
+    new URModel(
+        new ArrayList<Tuple2<String, IndexedDataset>>(),
+        pList,
+        getRankingMapping(),
+        false,
+        sc).save(dateNames, esIndex, esType);
+    return new NullModel();
+  }
+
+
+  @Override
+  public PredictedResult predict(NullModel model, Query query) {
+    logger.info("Query received, user id: ${query.user}, item id: ${query.item}");
+
+    List<String> queryEventNames = query.getEventNamesOrElse(modelEventNames);
```

Comment:
should be updating the field `queryEventNames`, instead of creating a new local var